### PR TITLE
[chore] fix some warnings and flaky smt calls

### PR DESCRIPTION
**Pull request by chdoc**

### CODE REVIEW

File: examples/MEE-CBC/RCPA_CMA.ec [hunk 1]
```diff
@@ -571,7 +571,7 @@ theory EtM.
           wp; sp. exists* ek{1}, p0{1}; elim* => _k _p.
           call (_: ={glob E, k, p} /\ k{1} = _k /\ p{1} = _p ==> ={glob E, res} /\ dec _k res{1} = Some _p).
             by conseq (_: ={glob E, k, p} ==> ={glob E, res}) (enc_sem _k _p); proc true.
-          by skip; smt.
+          skip => />; progress; smt(@FSet).
```
Comment (strub) [inline examples/MEE-CBC/RCPA_CMA.ec:574]: `progress` is deprecated. I am surprised that you need it after the use of `/>`

Comment (chdoc) [reply examples/MEE-CBC/RCPA_CMA.ec:574]: Yes, I was also surprised to see that neither `skip => />; smt.` nor `skip => />; smt(@FSet).` succeeds, given that the `/>` cuts down the size of the goal by about half.

Comment (fdupress) [reply examples/MEE-CBC/RCPA_CMA.ec:574]: This is likely one of those instances where simply introducing the memory or memories at the head of the conclusion speeds things up.

`progress` has the side effect of introducing them.

**Comment by chdoc** [comment]
Note that there is one warning that I didn't manage to fix:
```
[ϟ] examples/UC/RndO.ec: examples/UC/RndO.ec:237 global axiom sampleto_ll in section
[✓] [0030/0041] [ 0:06.1] examples/UC/RndO.ec
[✓] [0031/0041] [ 0:30.7] examples/UC/composition_cost.ec
[ϟ] examples/UC/dh_enc_cost.ec: examples/UC/dh_enc_cost.ec:2 global axiom sampleto_ll in section
```
Making the axiom global causes smt calls in `dh_enc_cost` to fail. :frowning:

File: examples/MEE-CBC/RCPA_CMA.ec (hunk 1 above)
Comment (chdoc) [reply examples/MEE-CBC/RCPA_CMA.ec:574]: In this case it also splits a large goal into several smaller ones, each presumably getting the full smt time allowance.

Comment (fdupress) [reply examples/MEE-CBC/RCPA_CMA.ec:574]: by skip=> />; smt(in_fsetU in_fset1).

**Comment by fdupress** [comment]
> Note that there is one warning that I didn't manage to fix:

Yes, I've already asked @mbbarbosa about it, and it's just... there. I think the libraries would need upgraded in ways that break other things to make more parameters local and better support that style of proofs.

**Comment by chdoc** [comment]
What is puzzling me is that moving the axiom from the section to just after the declaration of `sampleto`  changes nothing other than the order of the lemmas in the local `GenEager` theory of `dh_enc_cost` theory.  If this leads to smt failures, then those smt calls seem to be really fragile.

**Comment by oskgo** [comment]
Isn't `smt(@Theory)` almost as bad as `smt.`? They can both start timing out as lemmas are added to theories. I've had this issue with `smt(@Theory)` before where it started timing out in one of the examples in one of my PRs (probably the `DynMatrix` PR).

On a related note, would it be possible to have `smt` print which lemmas are actually used? I know that asking for the actual proof is a tall order, but maybe just the lemmas used is feasible?

**Comment by fdupress** [comment]
@oskgo This requires changes to Why3 that we don't have the capacity to implement. They suggested they'd take a PR, though if you want to take it on.

**Comment by chdoc** [comment]
> Isn't `smt(@Theory)` almost as bad as `smt.`? They can both start timing out as lemmas are added to theories. I've had this issue with `smt(@Theory)` before where it started timing out in one of the examples in one of my PRs (probably the `DynMatrix` PR).

I guess it's a compromise. One specific theory is still better than every lemma from the library that looks remotely related. So this should only break if lemmas are added to this particular theory.

**Comment by strub** [comment]
Moreover, not all SMT solvers have this ability. Once since that could be done (for a variant of `smt`) is to use Why3 to get the smtlib2 encoding of the goal and then call Z3 (e.g.) directly and use its ability to get the used lemmas.

**Comment by strub** [comment]
> > Isn't `smt(@Theory)` almost as bad as `smt.`? They can both start timing out as lemmas are added to theories. I've had this issue with `smt(@Theory)` before where it started timing out in one of the examples in one of my PRs (probably the `DynMatrix` PR).
> 
> I guess it's a compromise. One specific theory is still better than every lemma from the library that looks remotely related. So this should only break if lemmas are added to this particular theory.

Yes. At some point, we should craft good lemmas DB for the core theories.

**Pull request merged**